In [41]:
import pandas as pd
pd.options.display.max_rows = 200
pd.options.display.max_info_columns = 200

In [43]:
file = './dataset/train_data.csv'
df = pd.read_csv(file)

In [ ]:
df.head()

In [ ]:
df.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5531451 entries, 0 to 5531450
Data columns (total 190 columns):
 #    Column       Non-Null Count    Dtype  
---   ------       --------------    -----  
 0    customer_ID  5531451 non-null  object 
 1    S_2          5531451 non-null  object 
 2    P_2          5485466 non-null  float64
 3    D_39         5531451 non-null  float64
 4    B_1          5531451 non-null  float64
 5    B_2          5529435 non-null  float64
 6    R_1          5531451 non-null  float64
 7    S_3          4510907 non-null  float64
 8    D_41         5529435 non-null  float64
 9    B_3          5529435 non-null  float64
 10   D_42         791314 non-null   float64
 11   D_43         3873055 non-null  float64
 12   D_44         5257132 non-null  float64
 13   B_4          5531451 non-null  float64
 14   D_45         5529434 non-null  float64
 15   B_5          5531451 non-null  float64
 16   R_2          5531451 non-null  float64
 17   D_46         4319752 non-

In [44]:
N, D = df.shape

In [45]:
null_counts = df.isnull().sum()

In [46]:
null_percentages = null_counts / N * 100
has_NaN = null_percentages[null_percentages > 0]
print("No. of features with NaN values: ", has_NaN.size)
print(f"The feature with the most NaN is {has_NaN.idxmax()} ({has_NaN.max(): .2f}%)")
print(f"The feature with the most NaN is {has_NaN.idxmin()} ({has_NaN.min(): .10f}%)")
print(f"Features sorted by increasing NaN percentage:\n", has_NaN.sort_values(ascending=True))

No. of features with NaN values:  122
The feature with the most NaN is D_87 ( 99.93%)
The feature with the most NaN is R_7 ( 0.0000180784%)
Features sorted by increasing NaN percentage:
 R_7       0.000018
R_14      0.000018
B_40      0.000958
B_37      0.001012
R_12      0.001012
R_20      0.001356
B_6       0.004212
S_23      0.008045
S_26      0.011462
B_41      0.012474
D_109     0.028871
B_27      0.036446
B_2       0.036446
B_22      0.036446
B_20      0.036446
B_30      0.036446
B_19      0.036446
D_54      0.036446
B_16      0.036446
D_41      0.036446
B_33      0.036446
B_38      0.036446
B_3       0.036446
B_26      0.036446
D_45      0.036464
D_112     0.047908
B_25      0.125157
B_15      0.125157
S_25      0.232254
S_24      0.336132
S_22      0.343924
D_80      0.393622
D_74      0.393622
B_8       0.402571
D_72      0.428604
D_81      0.464381
D_52      0.534453
D_89      0.534453
D_84      0.534453
D_140     0.734563
D_102     0.734979
D_144     0.736281
D_133     0.772

In [55]:
NaN_summary = pd.DataFrame()
NaN_summary["Features"] = df.columns
NaN_summary["No. NaN"] = null_counts.values
NaN_summary["No. Data"] = [N] * D
NaN_summary["NaN Percent"] = null_percentages.values

In [63]:
NaN_summary.loc[NaN_summary["NaN Percent"] > 0].sort_values(by="NaN Percent")

Features  No. NaN  No. Data  NaN Percent
80       R_7        1   5531451     0.000018
100     R_14        1   5531451     0.000018
152     B_40       53   5531451     0.000958
142     B_37       56   5531451     0.001012
96      R_12       56   5531451     0.001012
117     R_20       75   5531451     0.001356
21       B_6      233   5531451     0.004212
131     S_23      445   5531451     0.008045
134     S_26      634   5531451     0.011462
171     B_41      690   5531451     0.012474
147    D_109     1597   5531451     0.028871
92      B_27     2016   5531451     0.036446
5        B_2     2016   5531451     0.036446
68      B_22     2016   5531451     0.036446
61      B_20     2016   5531451     0.036446
105     B_30     2016   5531451     0.036446
59      B_19     2016   5531451     0.036446
35      D_54     2016   5531451     0.036446
56      B_16     2016   5531451     0.036446
8       D_41     2016   5531451     0.036446
119     B_33     2016   5531451     0.036446
145     B_38     2016   5531451     0.036446
9        B_3     2016   5531451     0.036446
83      B_26     2016   5531451     0.036446
14      D_45     2017   5531451     0.036464
151    D_112     2650   5531451     0.047908
82      B_25     6923   5531451     0.125157
50      B_15     6923   5531451     0.125157
133     S_25    12847   5531451     0.232254
132     S_24    18593   5531451     0.336132
130     S_22    19024   5531451     0.343924
89      D_80    21773   5531451     0.393622
76      D_74    21773   5531451     0.393622
23       B_8    22268   5531451     0.402571
71      D_72    23708   5531451     0.428604
93      D_81    25687   5531451     0.464381
28      D_52    29563   5531451     0.534453
120     D_89    29563   5531451     0.534453
102     D_84    29563   5531451     0.534453
184    D_140    40632   5531451     0.734563
135    D_102    40655   5531451     0.734979
188    D_144    40727   5531451     0.736281
176    D_133    42716   5531451     0.772239
2        P_2    45985   5531451     0.831337
42      B_13    49519   5531451     0.895226
85      D_79    75939   5531451     1.372859
69      D_70    94917   5531451     1.715951
137    D_104   101548   5531451     1.835829
140    D_107   101548   5531451     1.835829
169    D_128   101548   5531451     1.835829
170    D_129   101548   5531451     1.835829
173    D_130   101548   5531451     1.835829
174    D_131   101548   5531451     1.835829
183    D_139   101548   5531451     1.835829
185    D_141   101548   5531451     1.835829
187    D_143   101548   5531451     1.835829
136    D_103   101548   5531451     1.835829
189    D_145   101548   5531451     1.835829
47      D_59   106725   5531451     1.929421
167    D_126   116816   5531451     2.111851
144     R_27   128703   5531451     2.326749
123     D_91   157216   5531451     2.842220
154    D_113   176716   5531451     3.194749
156    D_115   176716   5531451     3.194749
157    D_116   176716   5531451     3.194749
158    D_117   176716   5531451     3.194749
159    D_118   176716   5531451     3.194749
155    D_114   176716   5531451     3.194749
162    D_121   176716   5531451     3.194749
163    D_122   176716   5531451     3.194749
164    D_123   176716   5531451     3.194749
165    D_124   176716   5531451     3.194749
166    D_125   176716   5531451     3.194749
160    D_119   176716   5531451     3.194749
161    D_120   176716   5531451     3.194749
40      D_55   184803   5531451     3.340950
99      D_83   194473   5531451     3.515768
67      D_69   194473   5531451     3.515768
62      D_68   216503   5531451     3.914036
54      D_64   217442   5531451     3.931012
84      D_78   274319   5531451     4.959259
12      D_44   274319   5531451     4.959259
29       P_3   301492   5531451     5.450505
49      D_61   598052   5531451    10.811847
19      D_48   718725   5531451    12.993426
52      D_62   758161   5531451    13.706367
37       S_7  1020544   5531451    18.449843
7        S_3  1020544   5531451    18.449843

### Project Proposal Meeting
1. Data Understanding
    - there are 190 columns => 188 features (since the first column is index and the 2nd column is customer ID) 
    - 5,531,451 data points
    - 1 feature is the date. Could we elimiate it?
    - some feature has "NaN" value. How do we represent it? 
        run regression to predict it (statiscal analysis: modal, meadian, put the data on a distribution and see where the missing data would lie)
        0
        remove the columns
    
    => find a paper to read about how to deal with missing values

2. Papers:
    https://www.kaggle.com/competitions/amex-default-prediction/discussion/327135   

3. Team meeting for proposal recording on Thursday 10/6 at 5:30 PM

4. Trello Board: https://trello.com/b/F0tEjnUX/ml-project

5. Gantt Chart: https://docs.google.com/spreadsheets/d/1NwSPawBI_k9x3xHloXmnbROMbCaqwuFalB0XVgNrCJ8/edit?usp=sharing